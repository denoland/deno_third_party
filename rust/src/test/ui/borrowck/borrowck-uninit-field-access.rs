// revisions: ast mir
//[mir]compile-flags: -Z borrowck=mir

// Check that do not allow access to fields of uninitialized or moved
// structs.

#[derive(Default)]
struct Point {
    x: isize,
    y: isize,
}

#[derive(Default)]
struct Line {
    origin: Point,
    middle: Point,
    target: Point,
}

impl Line { fn consume(self) { } }

fn main() {
    let mut a: Point;
    let _ = a.x + 1; //[ast]~ ERROR use of possibly uninitialized variable: `a.x`
                     //[mir]~^ ERROR [E0381]

    let mut line1 = Line::default();
    let _moved = line1.origin;
    let _ = line1.origin.x + 1; //[ast]~ ERROR use of moved value: `line1.origin.x`
                                //[mir]~^ [E0382]

    let mut line2 = Line::default();
    let _moved = (line2.origin, line2.middle);
    line2.consume(); //[ast]~ ERROR use of partially moved value: `line2` [E0382]
                     //[mir]~^ [E0382]
}
