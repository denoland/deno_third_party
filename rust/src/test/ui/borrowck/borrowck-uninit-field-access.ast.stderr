error[E0381]: use of possibly uninitialized variable: `a.x`
  --> $DIR/borrowck-uninit-field-access.rs:24:13
   |
LL |     let _ = a.x + 1; //[ast]~ ERROR use of possibly uninitialized variable: `a.x`
   |             ^^^ use of possibly uninitialized `a.x`

error[E0382]: use of moved value: `line1.origin.x`
  --> $DIR/borrowck-uninit-field-access.rs:29:13
   |
LL |     let _moved = line1.origin;
   |         ------ value moved here
LL |     let _ = line1.origin.x + 1; //[ast]~ ERROR use of moved value: `line1.origin.x`
   |             ^^^^^^^^^^^^^^ value used here after move
   |
   = note: move occurs because `line1.origin` has type `Point`, which does not implement the `Copy` trait

error[E0382]: use of partially moved value: `line2`
  --> $DIR/borrowck-uninit-field-access.rs:34:5
   |
LL |     let _moved = (line2.origin, line2.middle);
   |                   ------------ value moved here
LL |     line2.consume(); //[ast]~ ERROR use of partially moved value: `line2` [E0382]
   |     ^^^^^ value used here after move
   |
   = note: move occurs because `line2.origin` has type `Point`, which does not implement the `Copy` trait

error: aborting due to 3 previous errors

Some errors occurred: E0381, E0382.
For more information about an error, try `rustc --explain E0381`.
